### PR TITLE
feat: crypto nonce randomness

### PR DESCRIPTION
**Pull request by krpeacock**
# Description
 
Currently default nonce generation is using Math.random, a pseudorandom source of randomness. We can take advantage of better tools when they are available

Fixes SDK-1194

# How Has This Been Tested?

Updated and new unit tests

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: packages/agent/src/agent/http/types.ts [hunk 1]
```diff
@@ -109,8 +110,9 @@ export function makeNonce(): Nonce {
   const buffer = new ArrayBuffer(16);
   const view = new DataView(buffer);
   const now = BigInt(+Date.now());
-  const randHi = Math.floor(Math.random() * 0xffffffff);
-  const randLo = Math.floor(Math.random() * 0xffffffff);
+  const randHi = Math.floor(randomNumber() * 0xffffffff); //?
+  const randLo = Math.floor(randomNumber() * 0xffffffff); //?
```
Comment (ericswanson-dfinity) [inline packages/agent/src/agent/http/types.ts:114]: Suggestion: make it so we don't (in several cases) divide by 4294967295 only to multiply by 0xffffffff here
```suggestion
  const randHi = Math.floor(randomNumber()); //?
  const randLo = Math.floor(randomNumber()); //?
```

File: packages/agent/src/utils/random.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+export const randomNumber = () => {
+  // determine whether browser crypto is available
+  if (typeof window !== 'undefined' && !!window.crypto && !!window.crypto.getRandomValues) {
+    const array = new Uint32Array(1);
+    window.crypto.getRandomValues(array);
+    return array[0] / 4294967295;
```
Comment (ericswanson-dfinity) [inline packages/agent/src/utils/random.ts:6]:
```suggestion
    return array[0];
```

File: packages/agent/src/utils/random.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  }
+  // A second check for webcrypto, in case it is loaded under global instead of window
+  if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
+    const array = new Uint32Array(1);
+    crypto.getRandomValues(array);
+    array;
+    array[0]; //?
+    return array[0] / 4_294_967_295;
```
Comment (ericswanson-dfinity) [inline packages/agent/src/utils/random.ts:14]:
```suggestion
    return array[0];
```

File: packages/agent/src/utils/random.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  type nodeCrypto = {
+    randomInt: (min: number, max: number) => number;
+  };
+
+  // determine whether node crypto is available
+  if (typeof crypto !== 'undefined' && (crypto as unknown as nodeCrypto).randomInt) {
+    return (crypto as unknown as nodeCrypto).randomInt(0, 1_000_000_000) / 4_294_967_295;
```
Comment (ericswanson-dfinity) [inline packages/agent/src/utils/random.ts:23]:
```suggestion
    return (crypto as unknown as nodeCrypto).randomInt(0, 1_000_000_000);
```

File: packages/agent/src/utils/random.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  // fall back to Math.random
+  return Math.random();
```
Comment (ericswanson-dfinity) [inline packages/agent/src/utils/random.ts:27]:
```suggestion
  return Math.random() * 0xffffffff;
```

File: packages/agent/src/utils/random.ts (hunk 4 above)
Comment (ericswanson-dfinity) [inline packages/agent/src/utils/random.ts:23]: Also is 1_000_000_000 correct here?

**Comment by krpeacock** [comment]
I was aiming initially to reproduce the behavior of Math.random, but I agree that yielding a random integer is preferable. I'll simplify, using your recommendations

File: packages/agent/src/utils/random.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+/**
+ * Generates a random unsigned 32-bit integer between 0 and 0xffffffff
+ * @returns {number} a random number
+ */
+export const randomNumber = () => {
```
Comment (ericswanson-dfinity) [inline packages/agent/src/utils/random.ts:5]: appropriate/useful?
```suggestion
export const randomNumber = (): number => {
```

File: packages/agent/src/utils/random.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+/**
```
Comment (ericswanson-dfinity) [inline packages/agent/src/utils/random.ts:1]: Could you write some rationale here for why we are not requiring some well-established crypto library and always using it here?

```
import { randomBytes } from 'crypto';

export const randomNumber = (): number => {
  return randomBytes(4).readUInt32BE(0);
};
```

File: packages/agent/src/utils/random.ts (hunk 6 above)
Comment (krpeacock) [reply packages/agent/src/utils/random.ts:5]: This is possibly the most controversial question in the TypeScript community 😁. I've chosen not to enforce return types for this codebase. 

Broadly speaking, this type is trivially inferred, so it does not add any utility for someone importing the function

File: packages/agent/src/utils/random.ts (hunk 7 above)
Comment (ericswanson-dfinity) [reply packages/agent/src/utils/random.ts:1]: Thanks!

**Pull request merged**